common/worker.c

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright © 2017 Drew Walters <drewwalters96@gmail.com>
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <error.h>
+#include <stddef.h>
+#include <worker.h>
+
+/*
+ * Queue of worker nodes.
+ */
+static struct worker_node *workers[MAX_WORKERS];
+
+void
+queue_worker(struct worker_node *worker)
+{
+	/* Add worker function to available queue index */
+	for (size_t i = 0; i < MAX_WORKERS; i++) {
+		if (!workers[i]) {
+			workers[i] = worker;
+			return;
+		}
+	}
+
+	panic("Worker function queue is full");
+}
+
+void
+process_workers(void)
+{
+	/* Call worker functions in queue */
+	for (size_t i = 0; i < MAX_WORKERS; i++) {
```

Comment:
Say we're working on `workers[1]`, so we've cleared `workers[0]`, and an interrupt comes in and the IRQ handler queues some work. It will go in `workers[0]`, but since we're already past that point, we will miss it and idle. So there needs to be another loop outside this one that keeps iterating through the array until the entire thing is empty.

Reply:
Done.